### ISSUE TITLE
since 3.10 my models train with a strong english accent

### ISSUE THREAD

**Issue by vertexgamer** [voicepaw/so-vits-svc-fork#438]
since 3.10 my models train with a strong english accent, i first thought it was an over training problem, but when training from scratch the same issue happen

**Comment by vertexgamer** [comment]
UPDATE: using infer on the same audio file with the same config and same model (which should work as it was trained before the issue appeared) seems to still have a strong english accent , suggesting that there is an issue in the "infer" process and not with the training process

**Comment by vertexgamer** [comment]
Ok, i just downgraded to 3.9.3 and everything works as expected.
It's CONFIRMED that newer versions have an accent bias.

If you are using a model in a language DIFFERENT than english, DO NOT USE newer versions!!!!

**Comment by liamlenholm** [comment]
@vertexgamer how do I downgrade?

**Comment by Lordmau5** [comment]
> Ok, i just downgraded to 3.9.3 and everything works as expected. It's CONFIRMED that newer versions have an accent bias.
> 
> If you are using a model in a language DIFFERENT than english, DO NOT USE newer versions!!!!

One thing I'm wondering now:

What if you train a model in 3.10+ and then infer on 3.9? Do you also get those accent results?

**Comment by vertexgamer** [comment]
> > Ok, i just downgraded to 3.9.3 and everything works as expected. It's CONFIRMED that newer versions have an accent bias.
> > If you are using a model in a language DIFFERENT than english, DO NOT USE newer versions!!!!
> 
> One thing I'm wondering now:
> 
> What if you train a model in 3.10+ and then infer on 3.9? Do you also get those accent results?

in my experience no, only the infer process affects the accent. But just to be sure, i'm training right now a model with 3.9.3. When i'm done i will come back with more info

**Comment by vertexgamer** [comment]
@Lordmau5 it seems that there is no hearable difference when using 3.9.3 models vs 3.10+ ones

**Comment by Lordmau5** [comment]
Hmm... okay that's interesting.

I know 3.10 did a switch from the `fairseq` to `transformers` library, which means it's one step less for building from what I can tell.
https://github.com/voicepaw/so-vits-svc-fork/commit/a2fe0f376d33f02987c91a57bd90a794de90a0e1

Apparently it's not relying on the pretrained contentvec model anymore and doesn't utilize it.
I saw another voice changer project that supports so-vits-svc models **did** require it though.

Maybe it has to do with that? @34j any thoughts? (Seeing as you made those changes)

**Comment by 34j** [comment]
I would like to suggest the possibility that the contents of final_proj are different

**Comment by 34j** [comment]
I'm not confident so anyone who has time please test it

**Comment by Lordmau5** [comment]
I did test one thing, and that was adding `"contentvec_final_proj": false` to the config, which ended up returning errors unfortunately...

What else should I try in regards to testing it? 🤔

**Comment by 34j** [comment]
That's not surprising because your model input is 256 channels and excepts final_projed input. You can download the model I trained from [here](https://huggingface.co/437aewuh/so-vits-svc-fork-kiritan/tree/main/ms-istft-768-44k) if you don't have one.

**Comment by Lordmau5** [comment]
> That's not surprising because your model input is 256 channels and excepts final_projed input.

Hmm, I just went with the template it gave me (started this model around a week ago and I couldn't spot any changes in regards to model input, ssl_dim or similar)

According to the wiki:
> The ssl_dim is the number of input channels, and the correct number of output channels for the officially trained ContentVec model is 768, but after applying final_proj it is 256.

Doesn't this mean that the config templates should be adjusted going forward?

___
I did try both of your models and they sound fine to me... Also no errors when inferring them

**Comment by 34j** [comment]
For now, I would like to suggest changing the default back to contentvec_final_proj=False and deal with it later.

**Comment by 34j** [comment]
Maybe so-vits-svc's contentvec is uniquely retrained in the final layer to enhance Japanese pronunciation? I don't know how it works, so I can't say for sure. ......

**Comment by Lordmau5** [comment]
Okay so, what I gathered just now:

Starting a new model and doing `svc pre-config` will select `so-vits-svc-4.0v1-legacy` by default.
Trying to set `"contentvec_final_proj": false` in that config file will return errors like mentioned above because it is a different model structure / config.

However, doing `svc pre-config -t so-vits-svc-4.0v1` will give the correct structure in which I can set the final_proj to true and then train with that.
It does seem like training that will take a bit longer however. Legacy at around 500 steps sounds better than that new one. That's fine though as long as it's mentioned.

I'm giving that training a go with the Kurzgesagt voice for testing to several thousand steps and report back.
___
The thing I see is that we need to figure out if a model is of `"type_": "hifi-gan"` or similar, as in **not legacy**, and in that case set it to use `contentvec_final_projc=False`.

Additionally, I've noticed a `n_speakers` variable that's set to 200 by default. I remember you saying something along the lines of "do we need 200 speakers?" and whether it could make the model smaller?

**Comment by 34j** [comment]
https://huggingface.co/lengyue233/content-vec-best

If you have more free time, you can follow this procedure to convert so-vits-svc's ContentVec and test it again.

**Comment by Lordmau5** [comment]
Yeah I converted the `checkpoint_best_legacy_500.pt` and loaded it in code instead of getting it from the `lengyue233` Hugging Face repository, the results are the same. It's still erroring... (Expecting 768 but providing 256 on a `so-vits-svc-4.0v1-legacy` model)

Trying to convert the non-legacy checkpoint is just erroring with that config (which makes sense)

**Comment by 34j** [comment]
What about with "contentvec_final_proj": false?

**Comment by 34j** [comment]
Note that final_proj is one nn.Linear that outputs 256 channels from 768 input channels.

**Comment by Lordmau5** [comment]
> What about with "contentvec_final_proj": true?

Yup, that works. But seeing as that's the default I assume we're back to square 1 with the English accent...

**Comment by 34j** [comment]
I'm not sure, but if the rebuilt version (for the purpose of replacing final_proj) still doesn't work, I think the only way is to include final_proj in the package or ask lengyue232 for help, or my code os wrong

**Comment by Lordmau5** [comment]
Modifying the `convert.py` script from lengyue233's repository a bit to remove the final_proj related code, I still get a config / model that has a `"hidden_size"` of 768. But we need one with 256.

I'm unsure how to convert it to a functional pytorch model...

**Comment by 34j** [comment]
> Modifying the convert.py script from lengyue233's repository a bit to remove the final_proj related code, I still get a config / model that has a "hidden_size" of 768. But we need one with 256.

You have better read our code first before talking, I'm saying that the weights for final proj is different between 2 original non Huggingface models and we need to replace it

**Comment by 34j** [comment]
> Note that final_proj is one nn.Linear that outputs 256 channels from 768 input channels.

**Comment by Lordmau5** [comment]
> You have better read our code first before talking, I'm saying that the weights for final proj is different between 2 original non Huggingface models and need to replace it

Aaaaah I see. I still don't understand much about the AI side of things with the project (I'm happy I can contribute with fixes here and there) so I apologize for that

**Comment by 34j** [comment]
I would like this to be resolved as soon as possible, do you have time now?

**Comment by 34j** [comment]
On the second thought, I think I'm the only person who can understand my dirty code and guess I should archive this repo

**Comment by 34j** [comment]
It is painful to be blamed for wasting the computing costs of the planet by having to train an incorrect model that was not identifiable for two days.

**Comment by 34j** [comment]
I've tried it and can't tell the difference......

**Comment by 34j** [comment]
3.10.0

https://user-images.githubusercontent.com/55338215/233765666-f6fa8669-632a-4296-a118-91c8cfb96d91.mp4

3.9.5

https://user-images.githubusercontent.com/55338215/233765714-784be9b2-23db-4a9d-b6db-78fa62ef2e91.mp4

The rebuilt one

https://user-images.githubusercontent.com/55338215/233765815-0279c011-101e-4e36-8ecd-7d2060498f9d.mp4

Still not fixed...

**Comment by 34j** [comment]
```
result1 = hubert(new_input, output_hidden_states=True)["hidden_states"][9]
result1 = hubert.final_proj(result1)
```

I didn't understand anything

**Comment by vertexgamer** [comment]
So have you guys fond the origin of the issue?

**Comment by Lordmau5** [comment]
> I would like this to be resolved as soon as possible, do you have time now?

I was unfortunately asleep at that time (7:25AM and I was awake until ilke 5AM hah), sorry :(

> It is painful to be blamed for wasting the computing costs of the planet by having to train an incorrect model that was not identifiable for two days.

I mean, you said it yourself before that you're still pretty new to this AI stuff if I remember correctly? Don't be too hard on yourself. There are bugs in other repositories that are way trickier to fix and might even go under the radar for longer 🙏 

> So have you guys fond the origin of the issue?

Well, 34j did push a fix in 3.10.5 for it - would you be able to give that another go and see if it is more comparable to what you got in 3.9?

**Comment by vertexgamer** [comment]
@Lordmau5 rn i'm training a model, when i finish i will try. I asked a friend to try it and it seems to be very similar to 3.9.3, but it might be placebo as the trained iterations are not the same
